artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -1175,6 +1179,17 @@ public void addConsumer(final Consumer consumer) throws Exception {
 
    }
 
+   @Override
+   public void addLingerRef(Consumer consumer, MessageReference lingerRef) throws Exception {
+      lingerRef.emptyConsumerID();
+      lingerRefMap.put(lingerRef, consumer);
```

Comment:
Beware putting `MessageReference` into a `Map`: `MessageReferenceImpl::hashCode` relies on `getMessage().hashCode()` and `CoreMessage::hashCode` will use the native hash code (like `java.lang.System#identityHashCode`).
It is known that it would break biased locking optimization on it (see https://blogs.oracle.com/dave/biased-locking-in-hotspot for more detail): given that `CoreMessage` heavily relies on `synchronized` operation, I would suggest to measure the perf impact of it first :+1:

Reply:
On that note, need to ensure order is preserved hashmap looses any ordering

Reply:
@franz1981 
> Beware putting `MessageReference` into a `Map`: `MessageReferenceImpl::hashCode` relies on `getMessage().hashCode()` and `CoreMessage::hashCode` will use the native hash code (like `java.lang.System#identityHashCode`).
> It is known that it would break biased locking optimization on it (see https://blogs.oracle.com/dave/biased-locking-in-hotspot for more detail): given that `CoreMessage` heavily relies on `synchronized` operation, I would suggest to measure the perf impact of it first
Thanks, franz. I'll have a look and measure the perf.

> On that note, need to ensure order is preserved hashmap looses any ordering
Good point. BTW, as i read from JMS spec, session is not thread safe, so closing consumer and committing session from which consumer is created can't be from multiple threads, correct? If not thread safe, the issue referred in this PR should be happening rarely? Just not sure whether it's deserved to fix this trivial problem by adding some tricky and fragile code.

Reply:
Remember Artemis is mutliprotocol. E.g. it supports more than just JMS

Reply:
Original order of message when rollback occurs needs to preserved so that theyre replayed the same order again

Reply:
Can use the new sorted methods instead of head.